### PR TITLE
init-flannel-network-health

### PR DESCRIPTION
**Pull request by calvix**
towards https://github.com/giantswarm/flannel-operator/issues/56

General idea:
this service will run in one pod together with `flanneld` and `k8s-network-bridge`.

`Flanneld` and `k8s-network-bridge` will get the specification for health checks and endpoints.

If some of the health checks fail then either `flanneld` or `k8s-network-bridge` is restarted and that should fix the error.

This can also catch issues when an interface is not properly cleaned and is present from old configuration.

The check itself is just looking at the network interface (either `br-xxxx` or `flannel.x`) and checking the ip of the interface.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+            wget -q $(curl -sS -H "Authorization: token $RELEASE_TOKEN" https://api.github.com/repos/giantswarm/architect/releases/latest | grep browser_download_url | head -n 1 | cut -d '"' -f 4)
+            chmod +x ./architect
+            ./architect version
+      - run:
+          command: ./architect build
+      - deploy:
+          command: |
+            if [ "${CIRCLE_BRANCH}" == "master" ]; then
+              ./architect deploy
+            fi
```
Comment (xh3b4sd) [inline .circleci/config.yml]: I would like to use a standard circle config like we do in other projects.

File: README.md [hunk 2]
```diff
@@ -1,14 +1,12 @@
 # flannel-network-health
 
-[![Build Status](https://api.travis-ci.org/giantswarm/example-opensource-repo.svg)](https://travis-ci.org/giantswarm/example-opensource-repo) [![Go Report Card](https://goreportcard.com/badge/github.com/giantswarm/example-opensource-repo)](https://goreportcard.com/report/github.com/giantswarm/example-opensource-repo) [![](https://godoc.org/github.com/giantswarm/example-opensource-repo?status.svg)](http://godoc.org/github.com/giantswarm/example-opensource-repo) [![](https://img.shields.io/docker/pulls/giantswarm/example-opensource-repo.svg)](http://hub.docker.com/giantswarm/example-opensource-repo) [![IRC Channel](https://img.shields.io/badge/irc-%23giantswarm-blue.svg)](https://kiwiirc.com/client/irc.freenode.net/#giantswarm)
```
Comment (xh3b4sd) [inline README.md]: I would like to keep at least the CI status badge.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	name        string = "flannel-network-health"
+	source      string = "https://github.com/giantswarm/flannel-network-health"
+)
+
+const (
+	ListenOn = ":8081"
+	MaxRetry = 100
+)
+
+func main() {
```
Comment (xh3b4sd) [inline main.go]: I would like to do the error handling of the main function like in the draughtsman-operator. There is a general issue somewhere to do this everywhere. See https://github.com/giantswarm/draughtsman-operator/blob/master/main.go#L35-L40 for the code.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	s.Logger = logger
+	if !s.LoadConfig() {
+		// failed to load config exiting
+		os.Exit(1)
+	}
+
+	// start blocking http server
+	http.HandleFunc("/bridge-healthz", s.CheckBridgeInterface)
+	http.HandleFunc("/flannel-healthz", s.CheckFlannelInterface)
+	http.ListenAndServe(ListenOn, nil)
```
Comment (xh3b4sd) [inline main.go]: I would like to use the microkit daemon server. This gives a metrics/version endpoint and the project/command structure as in other microservices. Also the signal handling is there and all the other magic. We have also an endpoint for health checks where we can implement health checks and execute custom logic. We should use this feature here as well.

File: server/server.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	FlannelInterface string
+	BridgeIP         string
+	FlannelIP        string
+}
+
+func DefaultConfig() *Server {
+	return &Server{}
+}
+
+func (s *Server) LoadConfig() bool {
```
Comment (xh3b4sd) [inline server/server.go]: I would like to have this business logic being implemented in a service package. This is now bound to the server which is only supposed to manage the network related protocols.

File: server/server.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	// load NIC interfaces from ENV
+	s.BridgeInterface = os.Getenv("NETWORK_BRIDGE_NAME")
+	s.FlannelInterface = os.Getenv("NETWORK_FLANNEL_DEVICE")
+	// read flannel file
+	fileContent, err := ioutil.ReadFile(os.Getenv("NETWORK_ENV_FILE_PATH"))
+	if err != nil {
+		s.Logger.Log(fmt.Printf("Error reading flannel file. %v", err.Error()))
+		return false
+	}
+	// get FLANNEL_SUBNET from flannel file via regexp
```
Comment (xh3b4sd) [inline server/server.go]: Above this code are arguments given that some special dependency provides. This dependency is in this case the OS. Below this code is the actual stateless business logic. I would like to separate both to be able to write tests for the actual algorithm.

File: server/server.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	checkBridge := exec.Command("ifconfig", s.BridgeInterface)
+	var output bytes.Buffer
+	checkBridge.Stdout = &output
+	err := checkBridge.Run()
+	if err != nil {
+		healthy = false
+		s.Logger.Log(fmt.Printf("Cant find bridge %s. %s", s.BridgeInterface, err.Error()))
+	}
+
+	if !strings.Contains(output.String(), s.BridgeIP) {
```
Comment (xh3b4sd) [inline server/server.go]: Same here towards testability.

File: main.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+		if _, err := os.Stat(flannelFile); !os.IsNotExist(err) {
+			break
+		}
+		logger.Log(fmt.Printf("Waiting for file %s to be created.", flannelFile))
+		time.Sleep(1 * time.Second)
+	}
+
+	s := server.DefaultConfig()
+	s.Logger = logger
+	if !s.LoadConfig() {
```
Comment (xh3b4sd) [inline main.go]: The naming should probably be something like `FlannelConfigExists`.

File: .circleci/config.yml (hunk 1 above)
Comment (r7vme) [reply .circleci/config.yml]: We already have project that uses Circle 2.0 and @JosephSalisbury was OK with it. 

https://github.com/giantswarm/k8s-kvm/blob/master/.circleci/config.yml

So i assume it's good to have new projects in 2.0

File: server/server.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+	s.BridgeIP = flannelIP.String()
+	// debug output
+	s.Logger.Log(fmt.Printf("Loaded Config: %+v", s))
+
+	return true
+}
+
+func (s *Server) CheckBridgeInterface(w http.ResponseWriter, r *http.Request) {
+	var healthy bool = true
+	checkBridge := exec.Command("ifconfig", s.BridgeInterface)
```
Comment (r7vme) [inline server/server.go]: from my side i can recommend to use iproute as native linux utility to manipulate networks.

e.g. `ip link show br-XXXX`

The best option can be to use https://github.com/vishvananda/netlink (which is used e.g. in calico)

**Comment by r7vme** [comment]
In general this service is useful if we want put more functionality, like expose some `/metrics` and or testing network connectivity too.

**Comment by calvix** [comment]
@r7vme Interesting, I didn't know about the live/readiness probe, but anyway I think we should use service that can give us easier checks and clear output.

**Comment by r7vme** [comment]
yes, as i said service will be beneficial, because we can put additional check and metrics there.

For example in Mirantis ex-colleagues did one for Calico. https://github.com/Mirantis/k8s-netchecker-server

Not 100% this use case, but idea is similar catch network (not only confifuration) problems. But this is future feature imo. as example sometimes i saw that flannel.XX interfaces stuck and i need to recreate to forward traffic.

File: server/server.go (hunk 9 above)
Comment (calvix) [reply server/server.go]: Thanks for the advice. migrated to `netlink` lib.

**Comment by calvix** [comment]
@xh3b4sd 
> How is that? As I see it the pod is then only restarted but the actually broken state is not cleaned, because we clean broken state with a specific job we have to schedule on demand. Or does re-creating the network bridge clean any broken state?

It doesnt fix it, but it will detect it and if pod is not ready we will get notiication via monitoring.

**Comment by calvix** [comment]
I am doing the migration to operator kit, it's quite complex for me so it takes time.

**Comment by xh3b4sd** [comment]
> I am doing the migration to operator kit, it's quite complex for me so it takes time.

I was thinking if we can separate some work. I understand the boilerplate magic coming from microkit takes time for somebody who is not familiar with the project structures. What would be important for me would be to separate the dependency related code and stateless algorithms of the actual health checks to be able to write tests. So in case you write tests and integrate this service into the KVM operator I can then do the microkit magic afterwards to get this going. WDYT? When you want to go through the microkit hustle because you want to "learn" it, go ahead and let me know when I can help.

**Comment by calvix** [comment]
I did the migration. I am not really convinced that I got all parts right :)  Can you check if you see some problems?

Some extra questions @xh3b4sd 
How can I configure which port will be used for the `healthz` endpoint? and how is the endpoint build? what url do I need to use to access it?

**Comment by calvix** [comment]
github is acting weird and doesn't wanna include new commits, hopefully it will be fixed soon

File: .circleci/config.yml (hunk 1 above)
Comment (calvix) [reply .circleci/config.yml]: yep this is circle ci config for v2

**Comment by xh3b4sd** [comment]
> How can I configure which port will be used for the healthz endpoint?

We do not configure any different port for health checks with microkit. The microendpoint health check we use is just `/healhtz`. There is only one port, which by default is `8000`. Same applies to metrics. It is the same port but `/metrics`. 

> and how is the endpoint build?

There is a common healthz endpoint from the microendpoint repo. See https://github.com/giantswarm/kubernetesd/blob/23bf5e6e78bd76051708024d8ff3a7c51dcd5cf6/server/endpoint/endpoint.go#L64-L77. This endpoint can be configured with healhtz services so we want to have something like https://github.com/giantswarm/kubernetesd/tree/23bf5e6e78bd76051708024d8ff3a7c51dcd5cf6/service/healthz but having two service packages inside of this healthz package where the separate health checks are implemented. 

> what url do I need to use to access it?

The default is `curl http://127.0.0.1:8000/healthz`.

> github is acting weird and doesn't wanna include new commits, hopefully it will be fixed soon

I noticed this in my own PRs as well. Ping me tomorrow again in case I did not respond to your code changes.

**Comment by calvix** [comment]
not sure how will that work with 2 separate endpoints

**Comment by calvix** [comment]
Does that mean that if I create 2 healthz checks then they are both executed when there is a request to `/healthz`? that's not so nice for this use case :(

I definitely need to be able to configure a port for the server as flannel-network components are running with `--net=host` and with that we cant have more than one flannel-network-health running on one machine because of the port overlap.

**Comment by xh3b4sd** [comment]
> not sure how will that work with 2 separate endpoints

There is one endpoint that has multiple services configured. 

> I definitely need to be able to configure a port for the server as flannel-network components are running with `--net=host`

How can we then define any port? There can be anything that conflicts with the port we chose. Is it possible that Kubernetes gives us a free port and we define one via a Kubernetes service manifest? So that we can map named ports.

**Comment by calvix** [comment]
> How can we then define any port? There can be anything that conflicts with the port we chose. Is it possible that Kubernetes gives us a free port and we define one via a Kubernetes service manifest? So that we can map named ports.

My thinking is that we use something similar like k8s does. Use some big port base like `21000` and then add there unique number in terms of flannel which is vni.

so the math for the port is `21000 + vni_num = port_to_listen_on`

**Comment by kopiczko** [comment]
> If some of the health checks fail then either flanneld or k8s-network-bridge is restarted and that should fix the error.

I think that should be flannel-operator responsibility. That should be checked and processed during reconciliation loop. Reason for that is I don't want two separate components operate on a single resource.

Exposing metrics in separate repo is OK, but I think this could also be the part of flannel-opearator.

**Comment by xh3b4sd** [comment]
The problem is that health checks and reconciliation have different intervals. It makes sense to do health checks more often. It is good enough to reconcile certain resources less often. In this case it makes sense to me to use Kubernetes primitives. IMO we should try to use the provider primitives where we can. See what we use in AWS and Azure. The operators are only delegating to the provider primitives. For KVM the "provider" primitives in this case are given by Kubernetes itself, so we should make use of it IMO.

**Comment by calvix** [comment]
finished, RFR @xh3b4sd

File: .circleci/config.yml (hunk 1 above)
Comment (xh3b4sd) [reply .circleci/config.yml]: I see. Did not notice this is the v2 thing. When this is how we are doing it in v2 fine with me.

File: README.md [hunk 10]
```diff
@@ -1,14 +1,12 @@
 # flannel-network-health
-
-[![Build Status](https://api.travis-ci.org/giantswarm/example-opensource-repo.svg)](https://travis-ci.org/giantswarm/example-opensource-repo) [![Go Report Card](https://goreportcard.com/badge/github.com/giantswarm/example-opensource-repo)](https://goreportcard.com/report/github.com/giantswarm/example-opensource-repo) [![](https://godoc.org/github.com/giantswarm/example-opensource-repo?status.svg)](http://godoc.org/github.com/giantswarm/example-opensource-repo) [![](https://img.shields.io/docker/pulls/giantswarm/example-opensource-repo.svg)](http://hub.docker.com/giantswarm/example-opensource-repo) [![IRC Channel](https://img.shields.io/badge/irc-%23giantswarm-blue.svg)](https://kiwiirc.com/client/irc.freenode.net/#giantswarm)
+[![CircleCI](https://circleci.com/gh/giantswarm/flannel-network-health/tree/master.svg?style=svg)](https://circleci.com/gh/giantswarm/flannel-network-health/tree/master)
```
Comment (xh3b4sd) [inline README.md]: I would like to have this above the title like in our other projects. Lets also use the style `shield` like in ther projects. See for instance https://raw.githubusercontent.com/giantswarm/kubernetesd/master/README.md?token=AAhvQV6b-P1saaVyJEH51K2ZvRuDabflks5Z8HCnwA%3D%3D.

File: main.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+)
+
+func main() {
+	err := mainWithError()
+	if err != nil {
+		panic(fmt.Sprintf("%#v\n", microerror.Mask(err)))
+	}
+}
+
+func mainWithError() error {
```
Comment (xh3b4sd) [inline main.go]: Nice.

File: main.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+	f           *flag.Flag = flag.New()
+	description string     = "Flannel-network-health serves as health endpoint for network configuration created by flannel-operator."
+	gitCommit   string     = "n/a"
+	name        string     = "flannel-network-health"
+	source      string     = "https://github.com/giantswarm/flannel-network-health"
+)
+
+const (
+	MaxRetry       = 100
+	ListenPortBase = 21000
```
Comment (xh3b4sd) [inline main.go]: I feel like these things should also be configurable like the other flags. Uncertain.

File: glide.yaml [hunk 13]
```diff
@@ -0,0 +1,3 @@
+package: github.com/giantswarm/flannel-network-health
+import:
+- package: github.com/giantswarm/micrologger
```
Comment (xh3b4sd) [inline glide.yaml]: There should also be a dependency on microkit and you should actually need to use your PR you created to fix the listen address issue.

File: main.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+			return microerror.Mask(err)
+		}
+	}
+
+	newCommand.CobraCommand().Execute()
+
+	return nil
+}
+
+func waitForFlannelFile(newLogger micrologger.Logger) error {
```
Comment (xh3b4sd) [inline main.go]: Why should we do this via service initialization? I feel like this is part of the health checks and they define if something is wrong or not.

File: main.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+		}
+		newLogger.Log("debug", fmt.Sprintf("Waiting for file '%s' to be created.", flannelFile))
+		time.Sleep(1 * time.Second)
+	}
+	// all good
+	return nil
+}
+
+func buildListenAddress() (string, error) {
+	vniStr := strings.Split(os.Getenv("NETWORK_FLANNEL_DEVICE"), ".")[1]
```
Comment (xh3b4sd) [inline main.go]: IMO the port should maybe be configured to the service and not made up here. The KVM operator knows all the details and should be source of truth for such details. It is also that we then probably can get rid of this wait check for the flannel file above which should not be done via service initialization. Rule of thumb should be that we should try to have as less as possible differences in the `main.go` files because the bootstrapping of our microservices should always be the same. We have the level of business logic abstraction in the service implementations and there all custom magic should happen.

File: server/endpoint/endpoint.go [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+	var err error
+
+	var healthzEndpoint *healthz.Endpoint
+	{
+		healthzConfig := healthz.DefaultConfig()
+		healthzConfig.Logger = config.Logger
+		healthzConfig.Services = []healthzservice.Service{
+			config.Service.Healthz.Bridge,
+			config.Service.Healthz.Flannel,
+		}
```
Comment (xh3b4sd) [inline server/endpoint/endpoint.go]: Super nice.

File: server/endpoint/endpoint.go [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	return newEndpoint, nil
+}
+
+// Endpoint is the endpoint collection.
+type Endpoint struct {
+	Healthz *healthz.Endpoint
+	Version *version.Endpoint
+}
```
Comment (xh3b4sd) [inline server/endpoint/endpoint.go]: I know you just copied this. Lets make Pawel happy though and put the struct above the ` New` constructor. Lets also make sure we do this from now on everywhere.

File: service/error.go [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+var invalidConfigError = microerror.New("invalid config")
+
+// IsInvalidConfig asserts invalidConfigError.
+func IsInvalidConfig(err error) bool {
+	return microerror.Cause(err) == invalidConfigError
+}
+
+var invalidFlannelFile = microerror.New("Error reading flannel file")
```
Comment (xh3b4sd) [inline service/error.go]: Nitpick, would be cool to keep the errors aligned with other projects. See https://github.com/giantswarm/giantswarm/wiki/go-style-guide#custom-errors. In this case the error variable should be called `invalidFlannelFileError`.  The error message should just be `invalid flannel file` then. When the error is actually used you can annotate more detailed information on a per case basis via `microerror.Maskf`.

File: service/flannel.go [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	// debug output
+	c.Logger.Log("debug", fmt.Sprintf("Loaded Config: %+v", c.Flag.Service.NetworkConfig))
+	return nil
+}
+
+// fetch ENVs values and read flannel file
+func (c *Config) fetchConfFromOS() ([]byte, error) {
```
Comment (xh3b4sd) [inline service/flannel.go]: I like this separation.

File: service/flannel_test.go [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+				return conf.Flag.Service.NetworkConfig, err
+			},
+			expectedConfig: network.Network{},
+			flannelFileContent: []byte(`FLANNEL_NETWORK=198.168.0.0/24
+FLANNEL_SUBNET=_x.68.c.0/30
+FLANNEL_MTU=1450
+FLANNEL_IPMASQ=false`),
+			expectedErr: invalidFlannelConfiguration,
+		},
+	}
```
Comment (xh3b4sd) [inline service/flannel_test.go]: Lets also have a test for an empty flannel file and one for flannel unrelated bullshit content. Just to make sure we do not panic or cause unexpected behaviour in the long run.

File: service/healthz/interfaceHealthz/key/key.go [hunk 21]
```diff
@@ -0,0 +1,9 @@
+package key
+
+import (
+	"github.com/vishvananda/netlink"
+)
+
+func GetInterfaceIP(ipList []netlink.Addr) string {
+	return ipList[0].IP.String()
```
Comment (xh3b4sd) [inline service/healthz/interfaceHealthz/key/key.go]: This code can panic. We should either handle errors or return an empty string in case an empty list is provided.

File: service/operator/service.go [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+func (s *Service) Boot() {
+	s.bootOnce.Do(func() {
+		// dummy  operator wait
+		s.logger.Log("debug", "Dummy operator boot.")
+		for {
+			time.Sleep(1 * time.Hour)
+		}
+
+	})
+}
```
Comment (xh3b4sd) [inline service/operator/service.go]: We probably do no even need all of this. It is actually only the server and health checks that we are interested in here and microkit is sorting this already.

File: service/healthz/healthz.go [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+
+// New creates a new configured healthz service.
+func New(config Config) (*Service, error) {
+	var err error
+
+	var bridgeService healthz.Service
+	{
+		bridgeServiceConfig := interfaceHealthz.DefaultConfig()
+		bridgeServiceConfig.NetworkInterface.Name = config.NetworkConfig.BridgeInterface
+		bridgeServiceConfig.NetworkInterface.IP = config.NetworkConfig.BridgeIP
```
Comment (xh3b4sd) [inline service/healthz/healthz.go]: I would say we can simply configure the name and IP separately without the bundling struct `NetworkInterface`. We can then also get rid of the `_interface` imports. Also FYI lets not have camel cased package names because this is not go idiomatic. See also https://blog.golang.org/package-names.

File: service/operator/service.go (hunk 22 above)
Comment (calvix) [reply service/operator/service.go]: So I can remove whole `operator` part? Or which parts/files can i remove?

Comment (kopiczko) [reply service/operator/service.go]: I don't see need for `operator` package at all.

Comment (xh3b4sd) [reply service/operator/service.go]: Yes, remove `service/operator` completely. No need for it here.

File: service/healthz/interfaceHealthz/key/key.go (hunk 21 above)
Comment (calvix) [reply service/healthz/interfaceHealthz/key/key.go]: fixed

File: service/healthz/healthz.go (hunk 23 above)
Comment (calvix) [reply service/healthz/healthz.go]: fixed

File: service/operator/service.go (hunk 22 above)
Comment (calvix) [reply service/operator/service.go]: Cool, fixed

File: server/endpoint/endpoint.go (hunk 17 above)
Comment (calvix) [reply server/endpoint/endpoint.go]: fixed

File: main.go (hunk 15 above)
Comment (calvix) [reply main.go]: true!

Comment (calvix) [reply main.go]: We cant really remove flannel file from equation because the information about network that is used on each machine is not information known to operators and its up to flannel

File: main.go (hunk 14 above)
Comment (calvix) [reply main.go]: removed

File: main.go (hunk 12 above)
Comment (calvix) [reply main.go]: since flannel-operator already have this value i will also add to it via ENV and remove the calculation magic from here

File: service/error.go (hunk 18 above)
Comment (calvix) [reply service/error.go]: fixed

**Comment by kopiczko** [comment]
Please remove `flannel-network-health` binary and add it to the `.gitignore`.

**Comment by calvix** [comment]
changes: 
* remove bundled struct NetworkInterface
* rename interfaceHealthz to flannel
* fix panic issue in GetInterfaceIP
* align error names
* remove 'operator' part from project
* move waitForFlannel from main to service load config and remove builder for listenAddress
* add extra tests
* add gitignore

File: service/flannel.go [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+	// load NIC interfaces from ENV
+	c.Flag.Service.NetworkConfig.BridgeInterface = os.Getenv("NETWORK_BRIDGE_NAME")
+	c.Flag.Service.NetworkConfig.FlannelInterface = os.Getenv("NETWORK_FLANNEL_DEVICE")
```
Comment (kopiczko) [inline service/flannel.go]: There are random places in the project where env vars are read. I'd be nice to move that to `mian.go`, so it's easy to see what vars can be passed to the program in a single place.

Comment (calvix) [reply service/flannel.go]: fixed

**Comment by calvix** [comment]
Any more comments?

File: service/flannel_test.go [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+dependencies:
+  override:
+    - |
+      wget -q $(curl -sS -H "Authorization: token $RELEASE_TOKEN" https://api.github.com/repos/giantswarm/architect/releases/latest | grep browser_download_url | head -n 1 | cut -d '"' -f 4)
+    - chmod +x ./architect
+    - ./architect version
+`),
+			expectedErr: invalidFlannelConfigurationError,
+		},
+	}
```
Comment (xh3b4sd) [inline service/flannel_test.go]: The tests are super nice. <3

File: service/flannel_test.go [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+
+	for index, test := range tests {
+		networkConfig, err := test.config(test.flannelFileContent)
+
+		if microerror.Cause(err) != microerror.Cause(test.expectedErr) {
+			t.Fatalf("%v: unexcepted error, expected %v but got %v", index, test.expectedErr, err)
```
Comment (xh3b4sd) [inline service/flannel_test.go]: Please use proper format string placeholders. `index` for instance is a integer and should be handled by `%d`. For errors and other arbitrary types I usually tend to use `%#v` to get the go representation of the error when printing it.

File: service/flannel_test.go (hunk 25 above)
Comment (calvix) [reply service/flannel_test.go]: thanks  :)

File: service/error.go [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+	return microerror.Cause(err) == invalidConfigError
+}
+
+var invalidFlannelFileError = microerror.New("invalid flannel file")
+
+func IsInvalidFlannelFile(err error) bool {
+	return microerror.Cause(err) == invalidFlannelFileError
+}
+
+var invalidFlannelConfigurationError = microerror.New("Unable to find FLANNEL_SUBNET in flannel file")
```
[the remark this answers is not in the document]
Comment (calvix) [reply service/error.go]: ok, hopefully i understand it right this time

**Comment by calvix** [comment]
Here we go :) fixed last requested changes.

File: service/error.go [hunk 28]
```diff
@@ -17,14 +17,14 @@ func IsInvalidFlannelFile(err error) bool {
...
+var invalidFlannelConfigurationError = microerror.New("invalid flannel configuration")
 
 func IsInvalidFlannelConfiguration(err error) bool {
 	return microerror.Cause(err) == invalidFlannelConfigurationError
 }
 
-var parsingFlannelSubnetError = microerror.New("parsing flannel file")
+var failedParsingFlannelSubnetError = microerror.New("failed parsing flannel file")
 
 func IsParsingFlannelSubnet(err error) bool {
```
Comment (xh3b4sd) [inline service/error.go]: The matcher has to be called `IsFailedParsingFlannelSubnet` in case the variable is called `failedParsingFlannelSubnetError`.

**Comment by calvix** [comment]
Thanks for all the advice and bearing with me!
